### PR TITLE
refactor(compiler): Add support for i18n expressions

### PR DESCRIPTION
**Pull request by mmalerba**
Adds support for i18n on templates and i18n expressions. See individual commit messages for more detail.

### CODE REVIEW

File: packages/compiler/src/template/pipeline/src/ingest.ts [hunk 1]
```diff
@@ -142,22 +148,35 @@ function ingestElement(unit: ViewCompilationUnit, element: t.Element): void {
...
   ingestNodes(unit, element.children);
 
-  unit.create.push(ir.createElementEndOp(id, element.endSourceSpan));
+  const endOp = ir.createElementEndOp(id, element.endSourceSpan);
+  unit.create.push(endOp);
+
+  // If there is an i18n message associated with this element, insert i18n start and end ops.
+  if (element.i18n instanceof i18n.Message) {
+    const i18nBlockId = unit.job.allocateXrefId();
+    ir.OpList.insertAfter<ir.CreateOp>(ir.createI18nStartOp(i18nBlockId, element.i18n), startOp);
```
Comment (dylhunn) [inline packages/compiler/src/template/pipeline/src/ingest.ts:166]: I think these probably should have source maps? Not necessary in this PR of course, but we'll need to figure this out from the tests

**Review by dylhunn — APPROVED** [review]

File: packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_i18n/multiple_pipes.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+consts: function() {
+  __i18nMsg__('{$interpolation} and {$interpolation_1}', [['interpolation', String.raw`\uFFFD0\uFFFD`], ['interpolation_1', String.raw`\uFFFD1\uFFFD`]], {original_code: {'interpolation': '{{ valueA | pipeA }}', 'interpolation_1': '{{ valueB | pipeB }}'}}, {})
+  __i18nMsg__('{$startTagSpan}{$interpolation}{$closeTagSpan} and {$interpolation_1} {$startTagSpan}and {$interpolation_2}{$closeTagSpan}', [['closeTagSpan', String.raw`[\uFFFD/#6\uFFFD|\uFFFD/#9\uFFFD]`], ['interpolation', String.raw`\uFFFD0\uFFFD`], ['interpolation_1', String.raw`\uFFFD1\uFFFD`], ['interpolation_2', String.raw`\uFFFD2\uFFFD`], ['startTagSpan', String.raw`[\uFFFD#6\uFFFD|\uFFFD#9\uFFFD]`]], {original_code: {'closeTagSpan': '</span>', 'interpolation': '{{ valueA | pipeA }}', 'interpolation_1': '{{ valueB | pipeB }}', 'interpolation_2': '{{ valueC | pipeC }}', 'startTagSpan': '<span>'}}, {})
+  …
```
Comment (mmalerba) [inline packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_i18n/multiple_pipes.js:4]: Note: there is an `i18npostprocess` instruction that is supposed to go here, but I haven't implemented support for it yet and this PR was getting big already, so I've ignored it for now

File: packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_i18n/multiple_pipes.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    $r3$.ɵɵpipe(7, "pipeA");
+    $r3$.ɵɵpipe(8, "pipeB");
+    $r3$.ɵɵelement(9, "span");
+    $r3$.ɵɵpipe(10, "pipeC");
+    $r3$.ɵɵi18nEnd();
+    $r3$.ɵɵelementEnd();
+  }
+  if (rf & 2) {
+    $r3$.ɵɵadvance(3);
+    $r3$.ɵɵi18nExp($r3$.ɵɵpipeBind1(2, …, ctx.valueA))($r3$.ɵɵpipeBind1(3, …, ctx.valueB));
```
Comment (mmalerba) [inline packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_i18n/multiple_pipes.js:25]: Note: the variable offsets for the pipes were incorrect. I'm not sure if this is a bug or not, but its something I've seen in other tests as well. I ignored it for now to focus on making sure we have the right instructions in the right order

**Comment by alxhub** [comment]
This PR was merged into the repository by commit 8d09e9e01380efd1f493139c9ae3eb4573e4e061.

**Pull request closed without merging**